### PR TITLE
Add unleash_context

### PR DESCRIPTION
**Pull request by ivarconr**
Started on the node companion of standardizing unleash strategies:

see java changes here:
https://github.com/Unleash/unleash-client-java/pull/33

TODO:
- [x] Implement common strategies
- [x] ApplicationHostnameStategy 
- [x] Automatically add all common context-bound strategies
- [x] Write some documentation on how to provide the UnleashContext and why it matters.
- [x] Do some testing before merging

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "typings": "./lib/index.d.ts",
   "scripts": {
     "lint": "eslint .",
-    "build": "tsc src/*.ts -d --strictNullChecks --sourceMap --outdir ./lib --lib ES6,ES5",
+    "build": "tsc src/*.ts src/**/*.ts -d --strictNullChecks --sourceMap --outdir ./lib --lib ES6,ES5",
```
Comment (ivarconr) [inline package.json]: @sveisvei any fancy ant-path expressions we can use instead here, this looks a bit verbose...?

**Comment by SimenB** [comment]
This closes #6?

File: src/context.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export  interface Context {
```
Comment (SimenB) [inline src/context.ts]: double space, whaaaaaa

File: src/strategy/gradual-rollout-random.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { Strategy } from '../strategy';
+import { Context } from '../context';
+
+export class GradualRolloutRandomStrategy extends Strategy {
+    constructor () {
+        super('gradualRolloutRandom');
+    }
+
+    isEnabled (parameters: any, context: Context) {
+        const percentage: number = +parameters.percentage;
```
Comment (SimenB) [inline src/strategy/gradual-rollout-random.ts]: `Number(parameters.percentage)` is much cleaner than `+parameters.percentage` IMO

File: src/strategy/gradual-rollout-session-id.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        super('gradualRolloutSessionId');
+    }
+
+    isEnabled (parameters: any, context: Context) {
+        const sessionId = context.sessionId;
+        if(!sessionId) {
+            return false;
+        }
+
+        const percentage = +parameters.percentage;
```
Comment (SimenB) [inline src/strategy/gradual-rollout-session-id.ts]: same here, I prefer `Number`. You might not, though 😁

**Comment by ivarconr** [comment]
Yes indeed!

File: src/strategy/gradual-rollout-user-id.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { Strategy } from '../strategy';
+import { Context } from '../context';
+import { normalizedValue } from './util';
+
+export class GradualRolloutUserIdStrategy extends Strategy {
```
Comment (SimenB) [inline src/strategy/gradual-rollout-user-id.ts]: Identical to `GradualRolloutSessionIdStrategy`, could they share some code?

File: src/strategy/user-with-id-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { Strategy } from '../strategy';
+import { Context } from '../context';
+
+export class UserWithIdStrategy extends Strategy {
```
Comment (SimenB) [inline src/strategy/user-with-id-strategy.ts]: Identical to `RemoteAddressStrategy`. Share code? 😀 

Or make it something more generic than `User` or `Address`

File: test/strategy/default-strategy.test.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import { DefaultStrategy } from '../../lib/strategy/default-strategy';
+
+test('default strategy should be enabled', (t) => {
+    const strategy = new DefaultStrategy();
+    t.true(strategy.isEnabled());
+});
+
+test('default strategy should have correct name', (t) => {
+    const strategy = new DefaultStrategy();
+    t.deepEqual(strategy.name, 'default');
```
Comment (SimenB) [inline test/strategy/default-strategy.test.js]: `t.true(strategy.name === 'default');`

File: src/strategy/gradual-rollout-random.ts (hunk 3 above)
Comment (ivarconr) [inline src/strategy/gradual-rollout-random.ts]: I'm old school, but never to late to change habit!

File: src/strategy/gradual-rollout-session-id.ts (hunk 4 above)
Comment (ivarconr) [inline src/strategy/gradual-rollout-session-id.ts]: ditto

File: src/strategy/gradual-rollout-user-id.ts (hunk 5 above)
Comment (ivarconr) [inline src/strategy/gradual-rollout-user-id.ts]: yes i thought about it, but i feel the abstraction costs more than the duplication. it's easier to read the code when it is explicit!

File: test/strategy/default-strategy.test.js (hunk 7 above)
Comment (ivarconr) [inline test/strategy/default-strategy.test.js]: yes!

**Pull request merged**